Add support for deploying IPv6/Dual-Stack hybrid mode testbed

hongliangl opened the pull request:
The commit enhances the KinD deployment script to support
creating an IPv6 or dual-stack KinD cluster for testing Antrea in
hybrid traffic mode.

antoninbas commented on ci/kind/kind-setup.sh:
```diff
@@ -232,8 +236,12 @@ function configure_extra_networks {
   networks=()
```
I don't think this function is related to0 hybrid mode?

hongliangl replied:
Removed the changes. Additionally, I found that it seems the option `--extra-network` of `ci/kind/test-e2e-kind.sh` introduced in #5416 is no longer used in current KinD tests. cc @jianjuns

antoninbas commented on ci/kind/kind-setup.sh:
```diff
@@ -154,8 +154,12 @@ function configure_networks {
-    docker network create -d bridge --subnet $s $network >/dev/null 2>&1
+    docker_network_args=("-d" "bridge" "--subnet" "$s")
+    if version_ge "$docker_version" "28.0.0"; then
+      docker_network_args+=("-o" "com.docker.network.bridge.gateway_mode_ipv4=nat-unprotected")
```
what about the IPv6 case?

hongliangl replied:
Added.

antoninbas reviewed (commented):
could you clarify why this is needed? AFAIK, the hybrid e2e tests have been working as expected in CI

hongliangl commented:
> could you clarify why this is needed? AFAIK, the hybrid e2e tests have been working as expected in CI

Yes, the hybrid e2e tests have been always working as expected with Github action. 

However, I found that when I running the test with the command `./ci/kind/test-e2e-kind.sh --encap-mode hybrid` on a local Linux testbed:

it will take a long time to collect information about K8s cluster. See the timestamps of the outputs:

```
2025/07/10 02:23:33 Test logs (if any) will be exported under the '/tmp/antrea-test-1956485518' directory
2025/07/10 02:23:33 Creating K8s ClientSet
2025/07/10 02:23:33 Collecting information about K8s cluster

2025/07/10 02:25:33 Pod IPv4 network: '10.244.0.0/16'
2025/07/10 02:25:33 Service IPv4 network: '10.96.0.0/16'
2025/07/10 02:25:33 Num nodes: 3
2025/07/10 02:25:33 Applying Antrea YAML
```

After applying Antrea YAML, the process will not move forward because one antrea-agent (on control-plane Node) is not ready for tests.

```
NAMESPACE            NAME                                         READY   STATUS    RESTARTS     AGE     IP           NODE                 NOMINATED NODE   READINESS GATES
kube-system          antrea-agent-9jg8g                           1/2     Running   0            105s    172.19.0.4   kind-control-plane   <none>           <none>
kube-system          antrea-agent-s9k2p                           2/2     Running   0            105s    20.20.20.2   kind-worker2         <none>           <none>
kube-system          antrea-agent-t9thv                           2/2     Running   0            105s    20.20.20.3   kind-worker          <none>           <none>
kube-system          antrea-controller-89c47b5f4-5jjm5            1/1     Running   0            105s    20.20.20.2   kind-worker2         <none>           <none>
kube-system          coredns-675756d478-9lb2c                     1/1     Running   0            6m2s    10.244.0.3   kind-control-plane   <none>           <none>
kube-system          coredns-675756d478-vrlwt                     1/1     Running   0            6m2s    10.244.0.2   kind-control-plane   <none>           <none>
kube-system          etcd-kind-control-plane                      1/1     Running   0            6m10s   172.19.0.4   kind-control-plane   <none>           <none>
kube-system          kube-apiserver-kind-control-plane            1/1     Running   0            6m10s   172.19.0.4   kind-control-plane   <none>           <none>
kube-system          kube-controller-manager-kind-control-plane   1/1     Running   0            6m10s   172.19.0.4   kind-control-plane   <none>           <none>
kube-system          kube-proxy-6rdlf                             1/1     Running   1 (6m ago)   6m3s    20.20.20.3   kind-worker          <none>           <none>
kube-system          kube-proxy-mz8qb                             1/1     Running   0            6m4s    172.19.0.4   kind-control-plane   <none>           <none>
kube-system          kube-proxy-qq75v                             1/1     Running   0            6m3s    20.20.20.2   kind-worker2         <none>           <none>
kube-system          kube-scheduler-kind-control-plane            1/1     Running   0            6m10s   172.19.0.4   kind-control-plane   <none>           <none>
local-path-storage   local-path-provisioner-7dc846544d-chn9b      1/1     Running   0            6m3s    10.244.1.3   kind-worker2         <none>           <none>
```

The process will exit:

```
2025/07/10 02:27:27 Error when deploying Antrea: error when waiting for antrea-agent rollout to complete - rc: 1 - stdout: Waiting for daemon set "antrea-agent" rollout to finish: 0 of 3 updated pods are available...
Waiting for daemon set "antrea-agent" rollout to finish: 1 of 3 updated pods are available...
Waiting for daemon set "antrea-agent" rollout to finish: 2 of 3 updated pods are available...
 - stderr: error: timed out waiting for the condition
 - err: <nil>
FAIL    antrea.io/antrea/test/e2e       234.253s
FAIL
Cleaning testbed
Deleting routes
Deleting cluster "kind" ...
Deleted nodes: ["kind-worker2" "kind-worker" "kind-control-plane"]
Deleting external servers
Deleting VLAN subnets
Deleted networks: antrea-0
```

Without the patch, I found there are two iptables rules that block all traffic that are not sourced from subnet gateway interface and destined for the worker Nodes:

```
*raw                                                                                                                                                                                                                                         
:PREROUTING ACCEPT [0:0]                                                                                                                                                                                                                     
:OUTPUT ACCEPT [0:0]                                                                                                                                                                                                                         
-A PREROUTING -d 127.0.0.1/32 ! -i lo -p tcp -m tcp --dport 40625 -j DROP                                                                                                                                                                    
-A PREROUTING -d 20.20.20.2/32 ! -i br-f9e290a9039e -j DROP                                                                                                                                                                                  
-A PREROUTING -d 20.20.20.3/32 ! -i br-f9e290a9039e -j DROP                                                                                                                                                                                  
COMMIT       
```

```
300: br-f9e290a9039e: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether be:e1:43:89:e2:bd brd ff:ff:ff:ff:ff:ff
    inet 20.20.20.1/24 brd 20.20.20.255 scope global br-f9e290a9039e
       valid_lft forever preferred_lft forever
    inet6 fe80::bce1:43ff:fe89:e2bd/64 scope link 
       valid_lft forever preferred_lft forever
```

antoninbas commented on ci/kind/kind-setup.sh:
```diff
@@ -152,10 +153,49 @@ function configure_networks {
   i=0
   networks=()
-  for s in $SUBNETS; do
+  bridge_id=$(docker network inspect kind -f {{.ID}})
```
how do the changes in this function relate to the `docker_run_with_host_net iptables -I DOCKER-USER -j ACCEPT` command that we already execute?

hongliangl replied:
You just reminded me. I didn't notice this. This command is sufficient to allow connections across bridges. I have removed the changes.

antoninbas commented on ci/kind/kind-setup.sh:
```diff
@@ -152,10 +153,49 @@ function configure_networks {
+    network_bridge_id=$(docker network inspect $network -f {{.ID}})
+    network_bridge_interface="br-${network_bridge_id:0:12}"
+    if [[ "$IP_FAMILY" != "ipv6" ]]; then
+      docker_run_with_host_net iptables -t filter -I FORWARD 1 -i $bridge_interface -o $network_bridge_interface -j ACCEPT
```
isn't this redundant with using `nat-unprotected`?

hongliangl commented on ci/kind/kind-setup.sh:
```diff
@@ -152,10 +155,35 @@ function configure_networks {
-    echo "creating network $network with $s"
-    docker network create -d bridge --subnet $s $network >/dev/null 2>&1
+    docker_network_args=("-d" "bridge")
+      # gateway_mode options based on Docker version
```
Fixed

hongliangl commented on ci/kind/kind-setup.sh:
```diff
@@ -152,10 +155,35 @@ function configure_networks {
+      for subnet in "${subnet_array[@]}"; do
+        docker_network_args+=("--subnet" "$subnet")
+      done
+    fi
```
Updated to this.
